src/compiler/preprocess/index.ts

```diff
@@ -222,98 +172,38 @@ export default async function preprocess(
 	const script = preprocessors.map(p => p.script).filter(Boolean);
 	const style = preprocessors.map(p => p.style).filter(Boolean);
 
-	// sourcemap_list is sorted in reverse order from last map (index 0) to first map (index -1)
-	// so we use sourcemap_list.unshift() to add new maps
-	// https://github.com/ampproject/remapping#multiple-transformations-of-a-file
-	const sourcemap_list: Array<DecodedSourceMap | RawSourceMap> = [];
-
-	// TODO keep track: what preprocessor generated what sourcemap? to make debugging easier = detect low-resolution sourcemaps in fn combine_mappings
+	const result = new PreprocessResult(source, filename);
 
-	for (const fn of markup) {
+	// TODO keep track: what preprocessor generated what sourcemap?
+	// to make debugging easier = detect low-resolution sourcemaps in fn combine_mappings
 
-		// run markup preprocessor
-		const processed = await fn({
-			content: source,
+	for (const process of markup) {
+		const processed = await process({
```

Comment:
possible to make this similar to `result.update_source(await process_tag('script', process, result));`, so it becomes:

```js
result.update_source(await process_markup(process, result));
```

Reply:
Yes, this is much nicer!